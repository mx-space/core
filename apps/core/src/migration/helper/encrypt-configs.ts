import { plainToInstance } from 'class-transformer'

import 'reflect-metadata'

import { ENCRYPT } from '~/app.config'
import { register } from '~/global/index.global'
import { generateDefaultConfig } from '~/modules/configs/configs.default'
import * as optionDtos from '~/modules/configs/configs.dto'
import { encryptObject } from '~/modules/configs/configs.encrypt.util'
import { IConfig, IConfigKeys } from '~/modules/configs/configs.interface'
import { getDatabaseConnection } from '~/utils/database.util'

console.log(ENCRYPT)

const allOptionKeys: Set<IConfigKeys> = new Set()
Object.entries(optionDtos).reduce((obj, [key, value]) => {
  const optionKey = (key.charAt(0).toLowerCase() +
    key.slice(1).replace(/Dto$/, '')) as IConfigKeys
  allOptionKeys.add(optionKey)
  return {
    ...obj,
    [`${optionKey}`]: value,
  }
}, {})

async function main() {
  await register()
  const connection = await getDatabaseConnection()
  const db = connection.db
  const configs: any[] = []
  await db
    .collection('options')
    .find()
    .forEach((current) => {
      configs.push(current)
    })

  const mergedConfig = generateDefaultConfig()
  configs.forEach((field) => {
    const name = field.name as keyof IConfig

    if (!allOptionKeys.has(name)) {
      return
    }

    const value = field.value
    mergedConfig[name] = { ...mergedConfig[name], ...value }
  })

  const encrypted = encryptObject(plainToInstance(IConfig as any, mergedConfig))

  for await (const [key, value] of Object.entries(encrypted)) {
    configs[key] = value
    await db.collection('options').updateOne(
      {
        name: key,
      },
      {
        $set: {
          value,
        },
      },
    )
  }

  await connection.close()
  process.exit(0)
}

main()
