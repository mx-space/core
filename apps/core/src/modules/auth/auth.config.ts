import { machineIdSync } from 'node-machine-id'
import type { ServerAuthConfig } from './auth.implement'

import { MongoDBAdapter } from '@mx-space/complied/auth'

import { API_VERSION } from '~/app.config'
import { SECURITY } from '~/app.config.test'
import { isDev } from '~/global/env.global'
import { getDatabaseConnection } from '~/utils/database.util'

import { AUTH_JS_USER_COLLECTION } from './auth.constant'

export const authConfig: ServerAuthConfig = {
  basePath: isDev ? '/auth' : `/api/v${API_VERSION}/auth`,
  secret: SECURITY.jwtSecret || machineIdSync(),
  callbacks: {
    redirect({ url }) {
      return url
    },
  },
  providers: [],
  experimental: {
    enableWebAuthn: true,
  },
  adapter: MongoDBAdapter(
    getDatabaseConnection().then((c) => c.getClient()),
    {
      collections: { Users: AUTH_JS_USER_COLLECTION },
    },
  ),
}
