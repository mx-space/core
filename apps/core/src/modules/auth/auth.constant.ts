export const AUTH_JS_USER_COLLECTION = 'readers'
export const AUTH_JS_ACCOUNT_COLLECTION = 'accounts'

export const AuthInstanceInjectKey = Symbol('AuthInstance')
