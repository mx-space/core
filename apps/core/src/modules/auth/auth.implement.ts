import { IncomingMessage } from 'node:http'
import { MongoClient } from 'mongodb'
import type { BetterAuthOptions } from 'better-auth'
import type { ServerResponse } from 'node:http'

import {
  APIError,
  bearer,
  betterAuth,
  jwt,
  mongodbAdapter,
  toNodeHandler,
} from '@mx-space/complied/auth'

import { API_VERSION, CROSS_DOMAIN, MONGO_DB } from '~/app.config'

import {
  AUTH_JS_ACCOUNT_COLLECTION,
  AUTH_JS_USER_COLLECTION,
} from './auth.constant'

const client = new MongoClient(MONGO_DB.customConnectionString || MONGO_DB.uri)

const db = client.db()

export async function CreateAuth(config: BetterAuthOptions['socialProviders']) {
  const auth = betterAuth({
    database: mongodbAdapter(db),
    socialProviders: config,
    basePath: isDev ? '/auth' : `/api/v${API_VERSION}/auth`,
    trustedOrigins: CROSS_DOMAIN.allowedOrigins.reduce(
      (acc: string[], origin: string) => {
        if (origin.startsWith('http')) {
          return [...acc, origin]
        }
        return [...acc, `https://${origin}`, `http://${origin}`]
      },
      [],
    ),
    account: {
      modelName: AUTH_JS_ACCOUNT_COLLECTION,
      accountLinking: {
        enabled: true,
        trustedProviders: ['google', 'github'],
      },
    },
    appName: 'mx-core',

    plugins: [
      jwt({
        jwt: {
          definePayload: async (user) => {
            const account = await db
              .collection(AUTH_JS_ACCOUNT_COLLECTION)
              .findOne({
                userId: user.id,
              })
            return {
              id: user.id,
              email: user.email,
              provider: account?.provider,
              providerAccountId: account?.providerAccountId,
            }
          },
        },
      }),
      bearer(),
    ],
    user: {
      modelName: AUTH_JS_USER_COLLECTION,
      additionalFields: {
        isOwner: {
          type: 'boolean',
          defaultValue: false,
          input: false,
        },
        handle: {
          type: 'string',
          defaultValue: '',
        },
      },
    },
  })

  const handler = async (req: IncomingMessage, res: ServerResponse) => {
    try {
      res.setHeader('access-control-allow-methods', 'GET, POST')
      res.setHeader('access-control-allow-headers', 'content-type')
      res.setHeader(
        'Access-Control-Allow-Origin',
        req.headers.origin || req.headers.referer || req.headers.host || '*',
      )
      res.setHeader('access-control-allow-credentials', 'true')
      return toNodeHandler(auth)(
        Object.assign(new IncomingMessage(req.socket), req, {
          url: req.originalUrl,
        }),
        res,
      )
    } catch (error) {
      console.error(error)
      // throw error
      res.end(error.message)
    }
  }

  return {
    handler,
    auth: {
      options: auth.options,
      api: {
        getSession(params: Parameters<typeof auth.api.getSession>[0]) {
          return auth.api.getSession(params)
        },
        getProviders() {
          return Object.keys(auth.options.socialProviders || {})
        },
        async listUserAccounts(
          params: Parameters<typeof auth.api.listUserAccounts>[0],
        ) {
          try {
            const result = await auth.api.listUserAccounts(params)
            return result
          } catch (error) {
            if (error instanceof APIError) {
              return null
            }
            throw error
          }
        },
      },
    },
  }
}
