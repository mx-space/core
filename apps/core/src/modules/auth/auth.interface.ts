import type { AdapterSession, AdapterUser } from '@mx-space/complied/auth'

export type SessionUser = AdapterSession & { user: AdapterUser }

declare module '@mx-space/complied/auth' {
  export interface AdapterUser {
    isOwner: boolean
  }
}
