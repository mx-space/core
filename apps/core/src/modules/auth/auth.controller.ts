import { Transform } from 'class-transformer'
import {
  IsDate,
  isMongoId,
  IsNotEmpty,
  IsOptional,
  IsString,
} from 'class-validator'

import {
  Body,
  Delete,
  Get,
  NotFoundException,
  Post,
  Query,
} from '@nestjs/common'
import { EventEmitter2 } from '@nestjs/event-emitter'

import { ApiController } from '~/common/decorators/api-controller.decorator'
import { Auth } from '~/common/decorators/auth.decorator'
import { EventBusEvents } from '~/constants/event-bus.constant'
import { MongoIdDto } from '~/shared/dto/id.dto'

import { AuthService } from './auth.service'

export class TokenDto {
  @IsDate()
  @IsOptional()
  @Transform(({ value: v }) => new Date(v))
  expired?: Date

  @IsString()
  @IsNotEmpty()
  name: string
}
@ApiController({
  path: 'auth',
})
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly eventEmitter: EventEmitter2,
  ) {}

  @Get('token')
  @Auth()
  async getOrVerifyToken(
    @Query('token') token?: string,
    @Query('id') id?: string,
  ) {
    if (typeof token === 'string') {
      return await this.authService
        .verifyCustomToken(token)
        .then(([isValid]) => isValid)
    }
    if (id && typeof id === 'string' && isMongoId(id)) {
      return await this.authService.getTokenSecret(id)
    }
    return await this.authService.getAllAccessToken()
  }

  @Post('token')
  @Auth()
  async generateToken(@Body() body: TokenDto) {
    const { expired, name } = body
    const token = await this.authService.generateAccessToken()
    const model = {
      expired,
      token,
      name,
    }
    await this.authService.saveToken(model)
    return model
  }
  @Delete('token')
  @Auth()
  async deleteToken(@Query() query: MongoIdDto) {
    const { id } = query
    const token = await this.authService
      .getAllAccessToken()
      .then((models) =>
        models.find((model) => {
          return (model as any).id === id
        }),
      )
      .then((model) => {
        return model?.token
      })

    if (!token) {
      throw new NotFoundException(`token ${id} is not found`)
    }
    await this.authService.deleteToken(id)

    this.eventEmitter.emit(EventBusEvents.TokenExpired, token)
    return 'OK'
  }
}
