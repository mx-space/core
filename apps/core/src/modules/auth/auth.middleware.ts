import type { NestMiddleware, OnModuleInit } from '@nestjs/common'
import type { BetterAuthOptions } from 'better-auth'
import type { IncomingMessage, ServerResponse } from 'node:http'

import { Inject } from '@nestjs/common'

import { EventBusEvents } from '~/constants/event-bus.constant'
import { SubPubBridgeService } from '~/processors/redis/subpub.service'

import { ConfigsService } from '../configs/configs.service'
import { AuthInstanceInjectKey } from './auth.constant'
import { CreateAuth } from './auth.implement'
import { InjectAuthInstance } from './auth.interface'

declare module 'http' {
  interface IncomingMessage {
    originalUrl: string
  }
}

export class AuthMiddleware implements NestMiddleware, OnModuleInit {
  private authHandler: Awaited<ReturnType<typeof CreateAuth>>['handler']

  constructor(
    private readonly redisSub: SubPubBridgeService,
    private readonly configService: ConfigsService,
    @Inject(AuthInstanceInjectKey)
    private readonly authInstance: InjectAuthInstance,
  ) {}

  async onModuleInit() {
    const handler = async () => {
      const oauth = await this.configService.get('oauth')

      const providers = {} as NonNullable<BetterAuthOptions['socialProviders']>
      oauth.providers.forEach((provider) => {
        if (!provider.enabled) return
        const type = provider.type as string

        const mergedConfig = {
          ...oauth.public[type],
          ...oauth.secrets[type],
        }
        switch (type) {
          case 'github': {
            if (!mergedConfig.clientId || !mergedConfig.clientSecret) return

            providers.github = {
              clientId: mergedConfig.clientId,
              clientSecret: mergedConfig.clientSecret,
            }
            break
          }

          case 'google': {
            if (!mergedConfig.clientId || !mergedConfig.clientSecret) return

            providers.google = {
              clientId: mergedConfig.clientId,
              clientSecret: mergedConfig.clientSecret,
            }

            break
          }
        }
      })
      const { handler, auth } = await CreateAuth(providers)
      this.authHandler = handler

      this.authInstance.set(auth)
    }
    this.redisSub.subscribe(EventBusEvents.OauthChanged, handler)
    await handler()
  }

  async use(req: IncomingMessage, res: ServerResponse, next: () => void) {
    if (!this.authHandler) {
      next()
      return
    }

    const bypassPath = ['/token', '/session', '/providers']

    if (bypassPath.some((path) => req.originalUrl.includes(path))) {
      next()
      return
    }
    if (req.method !== 'GET' && req.method !== 'POST') {
      next()
      return
    }

    return await this.authHandler(req, res)
  }
}
