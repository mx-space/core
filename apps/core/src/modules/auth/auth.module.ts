import { Global, Module } from '@nestjs/common'

import { AuthController } from './auth.controller'
import { AuthService } from './auth.service'

@Module({
  providers: [AuthService],
  controllers: [AuthController],
  exports: [AuthService],
})
@Global()
export class AuthModule {}
