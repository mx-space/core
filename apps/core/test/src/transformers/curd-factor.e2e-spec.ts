import { IsDefined, IsNumber } from 'class-validator'
import { eventEmitterProvider } from 'test/mock/processors/event.mock'

import { createE2EApp } from '@/helper/create-e2e-app'
import { authPassHeader } from '@/mock/guard/auth.guard'
import { modelOptions, prop, ReturnModelType } from '@typegoose/typegoose'

import { BaseModel } from '~/shared/model/base.model'
import { BaseCrudFactory } from '~/transformers/crud-factor.transformer'

@modelOptions({
  options: {
    customName: 'model-test',
  },
})
class TestModel extends BaseModel {
  @prop()
  @IsNumber()
  number: number

  @prop()
  @IsDefined()
  foo: string
}
export class TestController extends BaseCrudFactory({ model: TestModel }) {}

describe('BaseCrudFactory', () => {
  let testingModel: ReturnModelType<typeof TestModel>
  const proxy = createE2EApp({
    controllers: [TestController],
    providers: [...eventEmitterProvider],
    models: [TestModel],
    async pourData(modelMap) {
      const model = modelMap.get(TestModel)
      testingModel = model.model as any

      await model.model.create([
        {
          number: 1,
          foo: 'bar',
        },
      ])

      return async () => {
        return model.model.deleteMany({})
      }
    },
  })

  afterAll(() => {
    testingModel = null
  })

  test('GET /tests', async () => {
    const { app } = proxy
    const res = await app.inject({
      method: 'GET',
      url: '/tests',
    })
    const data = res.json()
    expect(res.statusCode).toBe(200)

    data.data.forEach((item) => {
      delete item.id
      delete item.created
    })
    expect(data).toMatchSnapshot()
  })

  test('POST /tests', async () => {
    const { app } = proxy
    const res = await app.inject({
      method: 'POST',
      url: '/tests',
      headers: {
        ...authPassHeader,
      },
      payload: {
        number: 2,
        foo: 'bar',
      },
    })
    const data = res.json()
    expect(res.statusCode).toBe(201)
    delete data.id
    delete data.created
    expect(data).toMatchSnapshot()
  })

  test('POST /tests should throw 422', async () => {
    const { app } = proxy
    const res = await app.inject({
      method: 'POST',
      url: '/tests',
      headers: {
        ...authPassHeader,
      },
      payload: {
        number: true,
      },
    })
    expect(res.statusCode).toBe(422)
  })

  test('PATCH /tests/:id', async () => {
    const { app } = proxy
    const docId = await testingModel
      .findOne()
      .lean()
      .then((doc) => doc?.id)

    const res = await app.inject({
      method: 'PATCH',
      url: `/tests/${docId}`,
      headers: {
        ...authPassHeader,
      },
      payload: {
        number: 3,
      },
    })

    expect(res.statusCode).toBe(204)

    const doc = await testingModel.findById(docId).lean()
    expect(doc.number).toBe(3)
  })

  test('DELETE /tests/:id', async () => {
    const { app } = proxy
    const docId = await testingModel
      .findOne()
      .lean()
      .then((doc) => doc?.id)

    const res = await app.inject({
      method: 'delete',
      url: `/tests/${docId}`,
      headers: {
        ...authPassHeader,
      },
    })

    expect(res.statusCode).toBe(204)

    const doc = await testingModel.countDocuments({
      id: docId,
    })
    expect(doc).toBe(0)
  })
})
