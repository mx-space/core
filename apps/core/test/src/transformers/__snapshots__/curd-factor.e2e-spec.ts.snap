// Vitest Snapshot v1

exports[`BaseCrudFactory > GET /tests 1`] = `
{
  "data": [
    {
      "foo": "bar",
      "number": 1,
    },
  ],
  "pagination": {
    "current_page": 1,
    "has_next_page": false,
    "has_prev_page": false,
    "size": 10,
    "total": 1,
    "total_page": 1,
  },
}
`;

exports[`BaseCrudFactory > POST /tests 1`] = `
{
  "foo": "bar",
  "number": 2,
}
`;
